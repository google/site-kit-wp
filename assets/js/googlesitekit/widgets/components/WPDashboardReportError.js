/**
 * WPDashboardReportError component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import { useLifecycles } from 'react-use';

/**
 * WordPress dependencies
 */
import { useInstanceId } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { CORE_UI } from '../../datastore/ui/constants';
import ReportError from '../../../components/ReportError';

export default function WPDashboardReportError( { moduleSlug, error } ) {
	const errorInstanceID = useInstanceId(
		WPDashboardReportError,
		'WPDashboardReportError'
	);

	const { setValue } = useDispatch( CORE_UI );

	const errorMessage = error.message;

	const selectInstanceID = useSelect( ( select ) =>
		select( CORE_UI ).getValue(
			`WPDashboardReportError-${ moduleSlug }-${ errorMessage }`
		)
	);

	useLifecycles(
		() => {
			setValue(
				`WPDashboardReportError-${ moduleSlug }-${ errorMessage }`,
				errorInstanceID
			);
		},
		() => {
			setValue(
				`WPDashboardReportError-${ moduleSlug }-${ errorMessage }`,
				undefined
			);
		}
	);

	if ( selectInstanceID !== errorInstanceID ) {
		return null;
	}

	return <ReportError moduleSlug={ moduleSlug } error={ error } />;
}

WPDashboardReportError.propTypes = {
	moduleSlug: PropTypes.string.isRequired,
	error: PropTypes.object.isRequired,
};
