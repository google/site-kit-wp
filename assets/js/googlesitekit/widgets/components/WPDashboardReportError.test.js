/**
 * WPDashboardReportError component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	createTestRegistry,
	provideModules,
	render,
} from '../../../../../tests/js/test-utils';
import WPDashboardReportError from './WPDashboardReportError';

describe( 'WPDashboardReportError', () => {
	it( 'should only render one error per module when there are multiple instances with same error', () => {
		const registry = createTestRegistry();

		const error = {
			code: 'test_error',
			message: 'Test error message',
			data: {
				reconnectURL: '#',
			},
		};

		function TestRender() {
			return (
				<div>
					<WPDashboardReportError
						moduleSlug="search-console"
						error={ error }
					/>
					<WPDashboardReportError
						moduleSlug="search-console"
						error={ error }
					/>
					<WPDashboardReportError
						moduleSlug="analytics-4"
						error={ error }
					/>
					<WPDashboardReportError
						moduleSlug="analytics-4"
						error={ error }
					/>
				</div>
			);
		}

		provideModules( registry );

		const { container } = render( <TestRender />, {
			registry,
		} );

		expect(
			container.querySelectorAll( '.googlesitekit-error-text' )
		).toHaveLength( 2 );

		expect( container ).toMatchSnapshot();
	} );
} );
