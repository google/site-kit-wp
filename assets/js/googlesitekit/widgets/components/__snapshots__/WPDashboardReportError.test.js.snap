// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WPDashboardReportError > should only render one error per module when there are multiple instances with same error 1`] = `
<div>
  <div>
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-typography googlesitekit-cta__title googlesitekit-typography--title googlesitekit-typography--small"
        >
          Data error in Search Console
        </h3>
        <div
          class="googlesitekit-cta__description"
        >
          <div
            class="googlesitekit-notice googlesitekit-notice--error"
          >
            <div
              class="googlesitekit-notice__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-notice__content"
            >
              <p
                class="googlesitekit-notice__description"
              >
                <span>
                  Error: Test error message (Please try again.)
                </span>
              </p>
            </div>
          </div>
        </div>
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div>
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="undefined?error_id=test_error"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-typography googlesitekit-cta__title googlesitekit-typography--title googlesitekit-typography--small"
        >
          Data error in Analytics
        </h3>
        <div
          class="googlesitekit-cta__description"
        >
          <div
            class="googlesitekit-notice googlesitekit-notice--error"
          >
            <div
              class="googlesitekit-notice__icon"
            >
              <svg />
            </div>
            <div
              class="googlesitekit-notice__content"
            >
              <p
                class="googlesitekit-notice__description"
              >
                <span>
                  Error: Test error message (Please try again.)
                </span>
              </p>
            </div>
          </div>
        </div>
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div>
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="undefined?error_id=test_error"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
